### PR TITLE
Add support for remoting.

### PR DESCRIPTION
**Pull request by mfoltzgoogle**
Re-pulling this branch.

***
<a href="https://pr-preview.s3.amazonaws.com/webscreens/openscreenprotocol/pull/192.html" title="Last updated on Sep 6, 2019, 4:21 PM UTC (c74434e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webscreens/openscreenprotocol/192/eaed041...c74434e.html" title="Last updated on Sep 6, 2019, 4:21 PM UTC (c74434e)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -975,10 +975,14 @@ the receiver decides to not attempt to load the resource at all.  If it chooses
 not to, it must respond with the appropriate failure result (such as timeout or
 invalid-url).  Additionally, the response must include the following:
 
-: state
+: state (optional)
```
Comment (mfoltzgoogle) [inline index.bs]: Why is the state property now optional?  The controller needs some data to maintain the local state of the media element.  That's true whether the media is flung by URL or streamed.

File: messages_appendix.cddl [hunk 2]
```diff
@@ -264,6 +264,14 @@ remote-playback-start-request = {
...
 }
 
+; type key 125
+remote-playback-start-response = {
+  response
+  1: &result ; result
+  ? 2: remote-playback-state ; state
+  ? 3: streaming-capabilities ; remoting-capabilities
+}
+p
```
Comment (mfoltzgoogle) [inline messages_appendix.cddl]: Looks like a typo.

File: index.bs [hunk 3]
```diff
@@ -1447,7 +1467,7 @@ audio encoding.
 Unlike most Open Screen Protocol messages, this one uses an
 array-based grouping rather than a struct-based grouping.  For
 required fields, this allows for a more efficient use of bytes on the
-wire, which is important for streaming audio because the payload is
+
```
Comment (mfoltzgoogle) [inline index.bs]: This looks like a mistake in the merge, which I'll fix in a commit :)

File: index.bs [hunk 4]
```diff
@@ -989,13 +993,29 @@ remote-playback-modify-request message with the following values:
...
 
+: remoting-audio
+:: Audio encodings that the controller can send to the receiver once they
+    are acked via the remote-playback-modify-response message.
+
+: remoting-video
+:: Videos encodings that the controller can send to the receiver once they
+    are acked via the remote-playback-modify-response message.
+
+: remoting-stats-interval
```
Comment (mfoltzgoogle) [inline index.bs]: Should this go in the remote-playback-start-request and remote-playback-start-response?
That is where the intervals are determined for a regular streaming session.

File: index.bs [hunk 5]
```diff
@@ -989,13 +993,29 @@ remote-playback-modify-request message with the following values:
...
+:: The frequency the contorller would like for the receiver to send
```
Comment (mfoltzgoogle) [inline index.bs]: Typo in controller

File: messages_appendix.cddl (hunk 2 above)
Comment (mfoltzgoogle) [reply messages_appendix.cddl]: Fixed

File: index.bs (hunk 3 above)
Comment (mfoltzgoogle) [reply index.bs]: Fixed

File: index.bs (hunk 5 above)
Comment (mfoltzgoogle) [reply index.bs]: Fixed

File: index.bs (hunk 1 above)
Comment (pthatcherg) [reply index.bs]: I don't remember.  I'll put it back :).

File: index.bs (hunk 4 above)
Comment (pthatcherg) [reply index.bs]: I moved it there.

**Comment by pthatcherg** [comment]
OK, I rebased on streaming and tried your idea.  See if you like it.

**Comment by pthatcherg** [comment]
OK, I thought about it some more and made a tweak to your idea that seems better (uses an optional parameter rather then embedding all the time).  I like how simple it ended up being.

File: index.bs [hunk 6]
```diff
@@ -478,6 +478,35 @@ those.  A request and a response includes a request ID which is an unsigned
 integer chosen by the requester.  Responses must include the request ID of the
 request they are associated with.
 
+Type Key Backwards Compatibility
```
Comment (mfoltzgoogle) [inline index.bs]: I think this is from a different PR.  Can you rebase?

Comment (mfoltzgoogle) [reply index.bs]: I guess it's my PR now so I'll do it :-)

**Comment by mfoltzgoogle** [comment]
This looks great.  I'll rebase and merge it.

We'll have to think about the entire lifecycle of the media element: specifically, what if we need to go from flinging to remoting and back again during the same remote playback.  Technically this is possible because you can activate MSE just by setting src= to a MediaSource.

**Pull request merged**